man/pamix.1
.\" this is the manpage of the pamix pulseaudio ncurses mixer
.TH pamix 1 "27 Nov 2024" "V" "pamix man page"

.SH DESCRIPTION
PAmix is a pavucontrol inspired ncurses based pulseaudio mixer

.SH CONFIGURATION
pamix is configured using a file called pamix.conf inside the $XDG_CONFIG_HOME or $HOME/.config, should it not be set.
.br
.start
.SH COMMANDS
.PP
PAmix conf files support the following commands:
.br
* bind

.SH bind
.PP
\fBSYNOPSIS:\fP bind KEYNAME MIXER\-COMMAND [ARGUMENT]

.PP
bind is used to bind a keyname to a mixer\-command.
.br
Some mixer\-commands require an argument.
.br
You can bind a keyname to multiple mixer\-commands.

.SH PAMIX\-COMMANDS
.PP
Pamix\-Commands can be bound to keys using the bind command and are used to interact with pamix.
.br
The following pamix\-commands are currently supported:

.SH quit
.PP
quit will cause PAmix to exit and takes no arguments.

.SH select\-tab
.PP
select\-tab will select one of the following tabs: Output Devices, Input Devices, Playback, Recording, Cards
.br
select\-tab takes the number of the tab to switch to starting at 0 in the order mentioned.

.SH select\-next and select\-prev
.PP
these commands are given the optional argument 'channel' they will select the next and previous channels.
if no argument is given they will select the next and previous entry in the displayed tab.

.SH set\-volume
.PP
this command takes the targetvalue in form of a double as an argument.
.br
depending on whether channels are locked, this command will set the volume of the selected entry/channel to the targetvalue given in the argument.
.br
\fIExample:\fP bind 0 set\-volume 1.0 \fI; this will set the volume to 100%\fP

.SH add\-volume
.PP
this command takes a deltavalue in form of a double as an argument.
.br
the deltavalue can be negative
\fIExample:\fP bind h add\-volume \-0.05 \fI; this will reduce the volume by 5%\fP

.SH cycle\-next and cycle\-prev
.PP
these commands will change the device or port of the currently selected entry.
.br
they dont take any arguments.

.SH toggle\-lock
.PP
this command toggles whether channels should be locked together for the currently selected entry
.br
and takes no arguments.

.SH toggle\-mute
.PP
toggles whether the currently selected entry is muted
.br
and takes no arguments.

.stop

.SH DEFAULT CONFIGURATION

Keybindings:
.br
F1      show Playback tab
.br
F2      show Recording tab
.br
F3      show Output devices tab
.br
F4      show Input devices tab
.br
F5      show Cards tab
.br
0-9     set volume to percentage (10%-100%)
.br
j/down  select next channel or entry
.br
k/up    select previous channel or entry
.br
h/left  decrease volume
.br
l/right increase volume
.br
c       un/lock channels
.br
s/S     select next/previous device/port



src/pamix.cpp
#include <pamix.hpp>

#include <unistd.h>
#include <configuration.hpp>
#include <condition_variable>
#include <queue>
#include <csignal>

// GLOBAL VARIABLES
Configuration configuration;

bool running = true;

// sync main and callback threads
std::mutex updMutex;
std::condition_variable cv;

std::queue<UpdateData> updateDataQ;

void quit()
{
	running = false;
	signal_update(false);
}

void __signal_update(bool all)
{
	{
		std::lock_guard<std::mutex> lk(updMutex);
		updateDataQ.push(UpdateData(all));
	}
	cv.notify_one();
}

void signal_update(bool all, bool threaded)
{
	if (threaded)
		std::thread(__signal_update, all).detach();
	else
		__signal_update(all);
}

void set_volume(pamix_ui *ui, double pct)
{
	auto it = ui->getSelectedEntryIterator();
	if (it != ui->m_Entries->end())
		it->second->setVolume(it->second->m_Lock ? -1 : (int)ui->m_SelectedChannel, (pa_volume_t)(PA_VOLUME_NORM * pct));
}

void add_volume(pamix_ui *ui, double pct)
{
	auto it = ui->getSelectedEntryIterator();
	if (it != ui->m_Entries->end())
		it->second->addVolume(it->second->m_Lock ? -1 : (int)ui->m_SelectedChannel, pct);
}

void cycle_switch(pamix_ui *ui, bool increment)
{
	auto it = ui->getSelectedEntryIterator();
	if (it != ui->m_Entries->end())
		it->second->cycleSwitch(increment);
}

void set_mute(pamix_ui *ui, bool mute)
{
	auto it = ui->getSelectedEntryIterator();
	if (it != ui->m_Entries->end())
		it->second->setMute(mute);
}

void toggle_mute(pamix_ui *ui)
{
	auto it = ui->getSelectedEntryIterator();
	if (it != ui->m_Entries->end())
		it->second->setMute(!it->second->m_Mute);
}

void set_lock(pamix_ui *ui, bool lock)
{
	auto it = ui->getSelectedEntryIterator();
	if (it != ui->m_Entries->end())
		it->second->m_Lock = lock;
}

void toggle_lock(pamix_ui *ui)
{
	auto it = ui->getSelectedEntryIterator();
	if (it != ui->m_Entries->end())
		it->second->m_Lock = !it->second->m_Lock;
}

void inputThread(pamix_ui *ui)
{
	while (running)
	{
		int ch = ui->getKeyInput();

#ifdef KEY_RESIZE
		if (ch == KEY_RESIZE)
		{
			endwin();
			refresh();
			signal_update(true);
		}
#endif

		bool isValidKey = ch != ERR && ch != KEY_RESIZE && ch != KEY_MOUSE;
		if (isValidKey) {
			configuration.pressKey(ch);
		}
		usleep(2000);
	}
}

void pai_subscription(PAInterface *, pai_subscription_type_t type) {
	bool updAll = (type & PAI_SUBSCRIPTION_MASK_INFO) != 0 || (type & PAI_SUBSCRIPTION_MASK_CONNECTION_STATUS) != 0;
	signal_update(updAll);
}

void sig_handle_resize(int) {
	endwin();
	refresh();
	signal_update(true, true);
}

void init_colors() {
	start_color();

	int background = use_default_colors() ? 0 : -1;

	init_pair(1, COLOR_GREEN, background);
	init_pair(2, COLOR_YELLOW, background);
	init_pair(3, COLOR_RED, background);
}

void sig_handle(int) {
	endwin();
}

void loadConfiguration() {
	char *home = getenv("HOME");
	char *xdg_config_home = getenv("XDG_CONFIG_HOME");

	std::string path;

	path = xdg_config_home ? xdg_config_home : std::string(home) += "/.config";
	path += "/pamix.conf";

	if (Configuration::loadFile(&configuration, path))
		return;

	char *xdg_config_dirs = getenv("XDG_CONFIG_DIRS");

	path = xdg_config_dirs ? xdg_config_dirs : "/etc/xdg";
	path += "/pamix.conf";
	size_t cpos = path.find(':');
	while (cpos != std::string::npos) {
		if (Configuration::loadFile(&configuration, path.substr(0, cpos)))
			return;
		path = path.substr(cpos + 1, path.length() - cpos - 1);
		cpos = path.find(':');
	}
	if (Configuration::loadFile(&configuration, path))
		return;

	// if config file cant be loaded, use default bindings
	configuration.bind("q", "quit");
	configuration.bind("KEY_F(1)", "select-tab", "2");
	configuration.bind("KEY_F(2)", "select-tab", "3");
	configuration.bind("KEY_F(3)", "select-tab", "0");
	configuration.bind("KEY_F(4)", "select-tab", "1");
	configuration.bind("KEY_F(5)", "select-tab", "4");

	configuration.bind("^I", "cycle-tab-next");

	configuration.bind("j", "select-next", "channel");
	configuration.bind("KEY_DOWN", "select-next", "channel");
	configuration.bind("J", "select-next");
	configuration.bind("k", "select-prev", "channel");
	configuration.bind("KEY_UP", "select-prev", "channel");
	configuration.bind("K", "select-prev");

	configuration.bind("h", "add-volume", "-0.05");
	configuration.bind("KEY_LEFT", "add-volume", "-0.05");
	configuration.bind("l", "add-volume", "0.05");
	configuration.bind("KEY_RIGHT", "add-volume", "0.05");

	configuration.bind("s", "cycle-next");
	configuration.bind("S", "cycle-prev");
	configuration.bind("c", "toggle-lock");
	configuration.bind("m", "toggle-mute");

	configuration.bind("1", "set-volume", "0.1");
	configuration.bind("2", "set-volume", "0.2");
	configuration.bind("3", "set-volume", "0.3");
	configuration.bind("4", "set-volume", "0.4");
	configuration.bind("5", "set-volume", "0.5");
	configuration.bind("6", "set-volume", "0.6");
	configuration.bind("7", "set-volume", "0.7");
	configuration.bind("8", "set-volume", "0.8");
	configuration.bind("9", "set-volume", "0.9");
	configuration.bind("0", "set-volume", "1.0");
}

void interfaceReconnectThread(PAInterface *interface) {
	while (running) {
		if (!interface->isConnected()) {
			interface->connect();
			signal_update(true);
		}
		sleep(5);
	}
}

void handleArguments(int argc, char **argv) {
	for (int i = 1; i < argc; i++) {
		std::string arg = argv[i];
		if (arg == "--version") {
			printf("PAmix v%s\n", GIT_VERSION);
			exit(0);
		}
	}
}

int main(int argc, char **argv) {
	handleArguments(argc, argv);
	loadConfiguration();

	setlocale(LC_ALL, "");
	initscr();
	init_colors();
	nodelay(stdscr, true);
	set_escdelay(25);
	curs_set(0);
	keypad(stdscr, true);
	meta(stdscr, true);
	noecho();

	signal(SIGABRT, sig_handle);
	signal(SIGSEGV, sig_handle);
#ifndef KEY_RESIZE
	signal(SIGWINCH, sig_handle_resize);
#endif

	PAInterface pai("pamix");
	pamix_ui pamixUi(&pai);
	if (configuration.has(CONFIGURATION_AUTOSPAWN_PULSE))
		pai.setAutospawn(configuration.getBool(CONFIGURATION_AUTOSPAWN_PULSE));

	entry_type initialEntryType = ENTRY_SINKINPUT;
	if (configuration.has(CONFIGURATION_DEFAULT_TAB)) {
		int value = configuration.getInt(CONFIGURATION_DEFAULT_TAB);
		if (value >= 0 && value < ENTRY_COUNT)
			initialEntryType = (entry_type)value;
	}

	pamixUi.selectEntries(initialEntryType);

	pai.subscribe(&pai_subscription);
	pai.connect();

	pamix_setup(&pamixUi);
	pamixUi.redrawAll();

	std::thread(&interfaceReconnectThread, &pai).detach();

	std::thread inputT(inputThread, &pamixUi);
	inputT.detach();

	while (running) {
		bool all;
		{
			std::unique_lock<std::mutex> lk(updMutex);
			cv.wait(lk, [] { return !updateDataQ.empty(); });
			if (updateDataQ.empty())
				continue;
			all = updateDataQ.front().redrawAll;
			updateDataQ.pop();
		}

		if (all)
			pamixUi.redrawAll();
		else
			pamixUi.redrawVolumeBars();
	}
	endwin();
	return 0;
}
